kafka-rest/src/main/java/io/confluent/kafkarest/controllers/SchemaManager.java

```diff
@@ -26,12 +27,42 @@
   /**
    * Returns the {@link RegisteredSchema schema} registered with the given {@code schemaId}.
    */
-  RegisteredSchema getSchemaById(EmbeddedFormat format, int schemaId);
+  RegisteredSchema getSchemaById(String subject, int schemaId);
 
   /**
-   * Parses and returns the potentially newly registered {@link RegisteredSchema schema}
-   * corresponding to {@code rawSchema}.
+   * Returns the {@link RegisteredSchema schema} registered with the given {@code schemaId}.
+   */
+  RegisteredSchema getSchemaById(
+      SubjectNameStrategy subjectNameStrategy, String topicName, boolean isKey, int schemaId);
+
+  /**
+   * Returns the {@link RegisteredSchema schema} registered with the given {@code schemaVersion}.
+   */
+  RegisteredSchema getSchemaByVersion(String subject, int schemaVersion);
+
+  /**
+   * Parses and returns {@code rawSchema}.
+   *
+   * <p>If the subject does not contain a registered schema equals to the parsed {@code rawSchema},
+   * one will be registered.
+   */
+  RegisteredSchema parseSchema(EmbeddedFormat format, String subject, String rawSchema);
```

Comment:
not into the name `parseSchema`, what about `registerOrLookupSchema`?

Reply:
Completely rewrote this bit. Ack.